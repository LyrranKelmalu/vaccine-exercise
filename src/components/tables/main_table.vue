<template>
  <v-card class="application_table">
    <h1>Table with orders</h1>
    <v-card-title>
      <v-text-field
        v-model="search"
        append-icon="mdi-magnify"
        label="Search"
        single-line
        hide-details
      ></v-text-field>  
    </v-card-title>
    <div class="cleanBtn">
      <v-btn
      :disabled="disabled"
      @click="cleanSearch()"
      depressed
      color="error"
      elevation="2"
      >Clean
      </v-btn>
    </div>
    <div v-if="!orders">
      <v-progress-circular
        class="text-center"
        :size="70"
        :width="7"
        color="primary"
        indeterminate
      ></v-progress-circular>
    </div>  
    <v-simple-table else dense>
      <template v-slot:default>
        <thead>
          <tr>
            <th class="text-left">
            </th>
            <th class="text-left">
              ID
            </th>
            <th class="text-left">
              Order Number
            </th>
            <th class="text-left">
              Responsible person
            </th>
            <th class="text-left">
              Health care district
            </th>
            <th class="text-left">
              Vaccine
            </th>
            <th class="text-left">
              Injections
            </th>
            <th class="text-left">
              Arrived
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            v-for="order in orders"
            :key="order.orderNumber"
          >
          <router-link
          :to="{ name: 'order', query: {id: order.id}}"
          >
            <v-btn
              color="primary"
              elevation="2"
              small
            >Open</v-btn>
          </router-link>  
            <td>{{ order.id }}</td>
            <td>{{ order.orderNumber }}</td>
            <td>{{ order.responsiblePerson }}</td>
            <td>{{ order.healthCareDistrict }}</td>
            <td>{{ order.vaccine }}</td>
            <td>{{ order.injections }}</td>
            <td>{{ order.arrived }}</td>
          </tr>
        </tbody>
      </template>
    </v-simple-table>
    <v-pagination
      v-model="page"
      :length="pageCount"
      :value="page"
      :total-visible="7"
    ></v-pagination>
  </v-card>
</template>

<script>
import {mapState, mapGetters, mapActions} from 'vuex'
import _ from 'lodash'

export default {
    data () {
      return {
        search: '',
        page: +this.$route.query.page || 1,
        pageSize: 15,
        pageCount: 0,
        orders: this.vaccineOrders,
        allOrders: [],
        disabled: true
      }
    },
    async mounted () {
      await this.fetchData()
      this.setupPagination(this.vaccineOrders)  
    },
    computed: {
    ...mapState(['vaccineOrders']),
    ...mapGetters(['sortByDate'])
    },
    methods: {
      ...mapActions(['fetchData']),
      setupPagination (array) {
        this.orders = array
        this.allOrders = _.chunk(this.orders, this.pageSize)
        this.pageCount = _.size(this.allOrders)
        this.orders = this.allOrders[this.page - 1] || this.allOrders[0]
      },
      currentPage () {
        this.$router.push(`${this.$route.path}?page=${this.page}`)
        this.orders = this.allOrders[this.page - 1] || this.allOrders[0]
      },
      cleanSearch () {
        this.search = ''
        this.disabled = true
      }
    },
    watch: {
      page: function () {
        this.currentPage()
      },
      search: function() {
        if (!this.search.length) {
          this.setupPagination(this.vaccineOrders)
        } else {
          const arr = this.sortByDate(this.search)
          this.setupPagination(arr)
          this.disabled = false
        }
      }
    }
}
</script>

<style>
.v-pagination__item--active {
    color: #fff;
    background-color: #1867c0 !important;
    border-color: #1867c0 !important;
}
.cleanBtn{
  padding-left: 16px;
  text-align: left;
}
</style>