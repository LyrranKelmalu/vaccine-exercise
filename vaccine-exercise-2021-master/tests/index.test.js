import 'jsdom-global/register'
import Vuex from 'vuex'
import Vue from 'vue'
import VueRouter from 'vue-router'
import { shallowMount, createLocalVue } from '@vue/test-utils'

import main_table from '../../src/components/tables/main_table.vue'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(VueRouter)
const router = new VueRouter()

Vue.config.ignoredElements = [
  'v-icon', 'v-card', 'v-card-text', 'v-pagination', 'v-simple-table',
  'v-progress-circular', 'v-btn', 'v-text-field', 'v-card-title'
]


describe('main_table.vue', () => {
  let actions
  let store
  let methods

  beforeEach(() => {
    actions = {
      fetchData: jest.fn()
    }

    methods = {
      setupPagination: jest.fn()
    }

    store = new Vuex.Store({
      actions
    })
  })

  it('check that component is mounted', () => {
    const wrapper = shallowMount(main_table, { store, router, localVue })
    expect(wrapper.is(main_table)).toBe(true)
  })

  it('renders correctly', () => {
    const wrapper = shallowMount(main_table, { store, router, localVue })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('dispatches "fetchData" when page is mounted', () => {
    const wrapper = shallowMount(main_table, { store, router, localVue })
    expect(actions.fetchData).toHaveBeenCalled()
  })

  it('setupPagination method is called when page is mounted', () => {
    const wrapper = shallowMount(main_table, { store, methods, router, localVue })
    expect(methods.setupPagination).toHaveBeenCalled()
  })
})