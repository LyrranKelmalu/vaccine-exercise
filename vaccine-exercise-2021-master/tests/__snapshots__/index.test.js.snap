// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main_table.vue renders correctly 1`] = `
<v-card
  class="application_table"
>
  <h1>
    Table with orders
  </h1>
   
  <v-card-title>
    <v-text-field
      append-icon="mdi-magnify"
      hide-details=""
      label="Search"
      single-line=""
    />
  </v-card-title>
   
  <div
    class="cleanBtn"
  >
    <v-btn
      color="error"
      depressed=""
      disabled="true"
      elevation="2"
    >
      Clean
    
    </v-btn>
  </div>
   
  <div>
    <v-progress-circular
      class="text-center"
      color="primary"
      indeterminate=""
      size="70"
      width="7"
    />
  </div>
   
  <v-simple-table
    dense=""
    else=""
  />
   
  <v-pagination
    length="0"
    total-visible="7"
    value="1"
  />
</v-card>
`;
